### PR TITLE
Integrate checkout signing API

### PR DESCRIPTION
**Pull request by robinandeer**
## Describe your changes

Integrate Checkout signing API on checkout / contact details page.

Move form logic inside checkout details pages.

Keep track of started "checkout signing" in cookie (key = checkoutId).

Flow: update contact details -> start sign -> poll for updates to checkout signing.

Still to be done:

- store the access token in a cookie after it's generated
- update cookie settings to be secure

## Justify why they are needed

Needed to support the "after sign" payment flow.

We stored the checkout signing ID in a cookie in case the user reloads the page while a signing is ongoing.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

[GRW-1562]: https://hedvig.atlassian.net/browse/GRW-1562?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by robinandeer** [comment]
Current dependencies on/for this PR:
* main
  * **PR #610** <a href="https://app.graphite.dev/github/pr/HedvigInsurance/racoon/610" target="_blank">Graphite</a> 
    * **PR #613** <a href="https://app.graphite.dev/github/pr/HedvigInsurance/racoon/613" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/HedvigInsurance/racoon/613?utm_source=stack-comment).

File: apps/store/src/components/CheckoutContactDetailsPage/CheckoutContactDetails.tsx [hunk 1]
```diff
@@ -2,46 +2,55 @@ import { Button, InputField, Space } from 'ui'
...
+            Continue to payment
+          </Button>
+        }
+      >
+        <Space y={1}>
+          <InputField
+            label="National identity number (DDMMYYXXXXX)"
+            name={FormElement.PersonalNumber}
+            required
+            defaultValue={prefilledData.personalNumber ?? undefined}
```
Comment (alebedev) [inline apps/store/src/components/CheckoutContactDetailsPage/CheckoutContactDetails.tsx:28]: Nit: Do you expect other false values in `prefilledData`? Missing key will return undefined anyway

File: apps/store/src/components/CheckoutContactDetailsPage/CheckoutSignPage.tsx [hunk 2]
```diff
@@ -2,50 +2,67 @@ import styled from '@emotion/styled'
...
+  checkoutId,
+  checkoutSigningId,
+  prefilledData,
+  onSuccess: onSignSuccess,
+}: CheckoutContactDetailsPageProps) => {
+  const [handleSubmit, loading] = useHandleSubmitContactDetailsAndSign({
+    checkoutId,
+    checkoutSigningId,
+    onSuccess() {
+      // @TODO: add access token to session
```
Comment (alebedev) [inline apps/store/src/components/CheckoutContactDetailsPage/CheckoutSignPage.tsx:17]: Are you planning to do this now or in separate PR?

File: apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  onSuccess: (accessToken: string) => void
+}
+
+export const useHandleSubmitContactDetailsAndSign = (params: Params) => {
+  const { checkoutId, checkoutSigningId: initialCheckoutSigningId, onSuccess } = params
+  const [checkoutSigningId, setCheckoutSigningId] = useState(initialCheckoutSigningId)
+
+  const { data } = useCheckoutSigningQuery({
+    skip: checkoutSigningId === null,
+    variables: checkoutSigningId ? { checkoutSigningId } : undefined,
```
Comment (alebedev) [inline apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:18]: Nit: `variables: { checkoutSigningId }`, we'll skip request if it's null anyway

File: apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        deleteCookie(checkoutId)
+      }
+    },
+  })
+
+  const [startSign, { loading: loadingStartSign }] = useCheckoutStartSignMutation({
+    variables: { checkoutId },
+    onCompleted(data) {
+      setCheckoutSigningId(data.checkoutStartSign.signing.id)
+      setCookie(checkoutId, data.checkoutStartSign.signing.id)
```
Comment (alebedev) [inline apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:35]: Nit: we use cookies as derived side effect here.  Maybe doing something like this would be cleaner?
```
// We need to backup checkoutSignId in a cookie to prevent loosing data
// if user reloads a page during signing
// - fetchCheckoutSigning on server side will see it
// and we'll be able to resume 
const useCheckoutSignIdCookie = (checkoutSigningId) => {
  useEffect(() => {
    if (checkoutSignId) {setCookie(...)} else {deleteCookie(...)}
  }, [checkoutSignId])
}

File: apps/store/src/pages/checkout/contact-details.tsx [hunk 5]
```diff
@@ -1,52 +1,64 @@
...
+  const locale = normalizeLocale(rawLocale)
+  if (!isLocale(locale)) return { notFound: true }
+
   try {
     const apolloClient = initializeApollo()
+
     const shopSession = await getCurrentShopSessionServerSide({ req, res, apolloClient })
+    const checkoutId = shopSession.checkout.id
+    if (shopSession.checkout.completedAt) {
+      return { redirect: { destination: PageLink.store({ locale }), permanent: false } }
```
Comment (alebedev) [inline apps/store/src/pages/checkout/contact-details.tsx:44]: Nit: add TODO about having receipt page in the future

**Review by alebedev — APPROVED** [review]
Looks good, some improvement ideas and questions in comments

File: apps/store/src/components/CheckoutContactDetailsPage/CheckoutContactDetails.tsx (hunk 1 above)
Comment (robinandeer) [reply apps/store/src/components/CheckoutContactDetailsPage/CheckoutContactDetails.tsx:28]: We pass values as `string | null` in `getServerSideProps` (`undefined` is not supported in JSON) -- hence this is needed.

Perhaps we can filter out keys where value is `null` instead.

File: apps/store/src/components/CheckoutContactDetailsPage/CheckoutSignPage.tsx (hunk 2 above)
Comment (robinandeer) [reply apps/store/src/components/CheckoutContactDetailsPage/CheckoutSignPage.tsx:17]: No, separate PR :)

File: apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts (hunk 3 above)
Comment (robinandeer) [reply apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:18]: I tried hehe but TS will complain if we do this since `checkoutSigningId = undefined` is invalid.

File: apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts (hunk 4 above)
Comment (robinandeer) [reply apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:35]: Hmm interesting approach -- I guess this could be considered "synchronize a component with an external system."

However, it seems like this falls under "[Sharing logic between event handlers](https://beta.reactjs.org/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers)" where React recommends not using `useEffect`.

Comment (alebedev) [reply apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:35]: Great article, did not see it before!

Comment (robinandeer) [reply apps/store/src/components/CheckoutContactDetailsPage/useHandleSubmitContactDetailsAndSign.ts:35]: Yeah the new react docs are really good!

**Review by guilhermespopolin — APPROVED** [review]

**Comment by robinandeer** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/HedvigInsurance/racoon/613?utm_source=stack-comment).


Job Reference: otDkpy8tB4rmLMp8y6We
